File: frontend/mf_pasand_app/lib/models/fund.dart
class Fund {
  final String schemeCode;
  final String schemeName;
  final String? isin;
  final String? fundHouse;
  final String? category;
  final String? fundType;
  final String? crisilRating;
  final double? expenseRatio;
  final double? aumCr;
  final double? returns1y;
  final double? returns3y;
  final double? returns5y;
  final double? volatility1y;
  final double? maxDrawdown1y;
  final double? minSip;
  final double? minLumpsum;
  final String? fundManager;
  final String? naturalText;
  final double? similarityScore;

  Fund({
    required this.schemeCode,
    required this.schemeName,
    this.isin,
    this.fundHouse,
    this.category,
    this.fundType,
    this.crisilRating,
    this.expenseRatio,
    this.aumCr,
    this.returns1y,
    this.returns3y,
    this.returns5y,
    this.volatility1y,
    this.maxDrawdown1y,
    this.minSip,
    this.minLumpsum,
    this.fundManager,
    this.naturalText,
    this.similarityScore,
  });

  factory Fund.fromJson(Map<String, dynamic> json) {
    return Fund(
      schemeCode: (json['scheme_code'] ?? json['schemeCode'] ?? '').toString(),
      schemeName: json['scheme_name'] ?? json['schemeName'] ?? 'Unknown Fund',
      isin: json['isin'] as String?,
      fundHouse: json['fund_house'] ?? json['fundHouse'] as String?,
      category: json['category'] as String?,
      fundType: json['fund_type'] ?? json['fundType'] as String?,
      crisilRating: json['crisil_rating'] ?? json['crisilRating'] as String?,
      expenseRatio: _toDouble(json['expense_ratio'] ?? json['expenseRatio']),
      aumCr: _toDouble(json['aum_cr'] ?? json['aumCr']),
      returns1y: _toDouble(json['returns_1y'] ?? json['returns1y']),
      returns3y: _toDouble(json['returns_3y'] ?? json['returns3y']),
      returns5y: _toDouble(json['returns_5y'] ?? json['returns5y']),
      volatility1y: _toDouble(json['volatility_1y'] ?? json['volatility1y']),
      maxDrawdown1y:
          _toDouble(json['max_drawdown_1y'] ?? json['maxDrawdown1y']),
      minSip: _toDouble(json['min_sip'] ?? json['minSip']),
      minLumpsum: _toDouble(json['min_lumpsum'] ?? json['minLumpsum']),
      fundManager: json['fund_manager'] ?? json['fundManager'] as String?,
      naturalText: json['natural_text'] ?? json['naturalText'] as String?,
      similarityScore:
          _toDouble(json['similarity_score'] ?? json['similarityScore']),
    );
  }

  static double? _toDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value);
    return null;
  }
}

File: backend/app/models/fund.py
"""Fund-related pydantic models."""

from typing import Optional

from pydantic import BaseModel


class FundBase(BaseModel):
    """Core identifying fields for a mutual fund scheme."""

    scheme_code: int
    scheme_name: str
    isin: Optional[str] = None
    fund_house: Optional[str] = None
    category: Optional[str] = None
    fund_type: Optional[str] = None


class FundMetrics(BaseModel):
    """Performance and risk metrics computed from NAV history."""

    returns_1y: Optional[float] = None
    returns_3y: Optional[float] = None
    returns_5y: Optional[float] = None
    volatility_1y: Optional[float] = None
    max_drawdown_1y: Optional[float] = None


class FundDetail(FundBase, FundMetrics):
    """Full fund record including metadata and computed metrics."""

    crisil_rating: Optional[str] = None
    expense_ratio: Optional[float] = None
    aum_cr: Optional[float] = None
    min_sip: Optional[float] = None
    min_lumpsum: Optional[float] = None
    fund_manager: Optional[str] = None
    natural_text: Optional[str] = None


class FundRecommendation(FundDetail):
    """Fund detail augmented with a similarity score from vector search."""

    similarity_score: float = 0.0


class FundListResponse(BaseModel):
    """Paginated list of funds."""

    funds: list[FundDetail]
    total: int
    page: int
    page_size: int
